core/src/main/java/fj/data/Eval.java

```diff
@@ -0,0 +1,209 @@
+   *
+   * @param a the supplier that evaluates a value.
+   * @return a lazy evaluation.
+   */
+  public static <A> Eval<A> later(F0<A> a) {
+    return new Later<>(a);
+  }
+
+  /**
+   * Constructs a lazy evaluation without caching.
+   *
+   * @param a the supplier that evaluates a value.
+   * @return a lazy evaluation.
+   */
+  public static <A> Eval<A> always(F0<A> a) {
+    return new Always<>(a);
+  }
+
+  /**
+   * Evaluates the computation and return its result.
+   * Depending on whether the current instance is lazy or eager the
+   * computation may or may not happen at this point.
+   *
+   * @return a result of this computation.
+   */
+  public abstract A value();
+
+  /**
+   * Transforms <code>Eval<A></code> into a <code>Eval<B></code> using
+   * the given function.
+   *
+   * Note: the computation of the given transformation is always lazy,
+   * even if it invoked for an eager <code>Now</code> instance. This computation
+   * is performed in O(1) stack space.
+   *
+   * @param f the transformation function.
+   * @return a transformed evaluation.
+   */
+  public <B> Eval<B> map(final F<A, B> f) {
+    return bind(a -> now(f.f(a)));
+  }
+
+  /**
+   * Alias for {@link #bind(F)}.
+   */
+  public <B> Eval<B> flatMap(final F<A, Eval<B>> f) {
+    return bind(f);
+  }
+
+  /**
+   * Transforms <code>Eval<A></code> into a <code>Eval<B></code> using
+   * the given function that directly produces <code>Eval<B></code>.
+   *
+   * Note: the computation of the given transformation is always lazy,
+   * even if it invoked for an eager <code>Now</code> instance. This computation
+   * is performed in O(1) stack space.
+   *
+   * @param f the transformation function.
+   * @return a transformed evaluation.
+   */
+  public <B> Eval<B> bind(final F<A, Eval<B>> f) {
+    return bindTrampoline(f, asTrampoline());
+  }
+
+  /**
+   * Transforms the current instance into a trampoline instance.
+   */
+  TrampolineEval<A> asTrampoline() {
+    return pureTrampoline(this);
+  }
+
+  /**
+   * Represents an eager computation.
+   */
+  private static final class Now<A> extends Eval<A> {
+    private final A a;
+
+    private Now(A a) {
+      this.a = a;
+    }
+
+    @Override
+    public A value() {
+      return a;
+    }
+  }
+
+  /**
+   * Represents a lazy computation that is evaluated only once.
+   */
+  private static final class Later<A> extends Eval<A> {
+    private final P1<A> memo;
+
+    private Later(F0<A> producer) {
+      this.memo = P.hardMemo(producer);
+    }
+
+    @Override
+    public A value() {
+      return memo._1();
+    }
+  }
+
+  /**
+   * Represents a lazy computation that is evaluated every time when it's requested.
+   */
+  private static final class Always<A> extends Eval<A> {
+    private final F0<A> supplier;
+
+    private Always(F0<A> supplier) {
+      this.supplier = supplier;
+    }
+
+    @Override
+    public A value() {
+      return supplier != null ? supplier.f() : null;
```

Comment:
should just be `return supplier.f()`.
In FJ we do not special on case `null` imput value: protecting each input parameter against null would litter the code with garbage, maybe also impact perf and, in this case, hide root issue by propagating a null value.

Only dedicated APIs like `Option.fromNull` explicitly deal with `null`.